keps/sig-network/3453-minimize-iptables-restore/README.md

```diff
@@ -0,0 +1,468 @@
+_any_ rule in a given iptables chain, then you must specify _every_
+rule in that chain. However, if you want to leave a chain completely
+unchanged, then you can just fail to mention any rules in that chain,
+and it will be left untouched. (If you want to delete a chain you have
+to explicitly request that.)
+
+Thus, [the proposed PR] keeps track of which Service and EndpointSlice
+objects have changed since the last `iptables-restore`, and only
+outputs the Service-specific chains for a Service (the `KUBE-SVC-*`,
+`KUBE-SVL-*`, `KUBE-EXT-*`, `KUBE-FW-*`, and `KUBE-SEP-*` chains) if
+the Service or its EndpointSlice(s) changed since the last
+`iptables-restore`. (The `KUBE-SERVICES`, `KUBE-EXTERNAL-SERVICES`,
+and `KUBE-NODEPORTS` chains are written in their entirety on every
+sync.)
+
+In theory this could be optimized further; eg, there are situations
+where only `KUBE-FW-*` changes, or where we need to rewrite the
+`KUBE-SERVICES` rules for a service but not any of the
+service-specific chains, or vice versa. However, these other
+optimizations would be more complicated, and would only result in a
+small additional performance improvement relative to the main
+improvement of not rewriting _every_ service and endpoint rule on
+_every_ resync.
+
+In the event that a partial `iptables-restore` call fails, kube-proxy
+will queue another resync, and fall back to doing a full
+`iptables-restore` that time. (Note that there are no expected cases
+where a partial restore would fail but a full restore would succeed;
+this would always indicate a bug in the code.) Likewise, it always
+does a full resync when there are topology-related changes to Node
+labels, and it always does a full resync at least once every
+`iptablesSyncPeriod`.
+
+[the proposed PR]: https://github.com/kubernetes/kubernetes/pull/110268
+
+### Risks and Mitigations
+
+#### Catastrophic Failure
+
+In the event of bugs in the code, kube-proxy could fail
+catastrophically, entirely breaking the cluster. This would be "bad".
+
+#### Subtle Optimization Failure
+
+The code could also fail more subtly. In the worst case, every attempt
+to do a partial restore might fail, causing the code to fall back
+to the slower full `iptables-restore` on the next sync, with the end
+result that kube-proxy becomes _less_ efficient than the old code that
+would have just done a full restore the first time.
+
+Less obviously, the new code might reliably cause a fail-and-retry
+only in certain circumstances (eg, when a service goes from 1 endpoint
+to 0, or when a particular Service property changes).
+
+Since, as mentioned above, there are no _expected_ cases where a
+partial restore will fail, we could notice that this failure mode was
+happening by having a metric for how often a partial restore fails,
+and if we find that it is happening (in e2e tests or real clusters),
+we can then debug or revert the bad code.
+
+#### Subtle Synchronization Delays
```

Comment:
Any interest in changing the settings for bounded frequency runner? Or updating it to account for the fact that many syncs are expected to be much cheaper?